store/bigtable/bigtable.go

```diff
@@ -0,0 +1,471 @@
+	btblChunksPerRow = stats.NewMeter32("store.bigtable.chunks_per_row", false)
+	// metric store.bigtable.rows_per_response is how many rows come per get response
+	btblRowsPerResponse = stats.NewMeter32("store.bigtable.rows_per_response", false)
+
+	// metric store.bigtable.chunk_operations.save_ok is counter of successful saves
+	chunkSaveOk = stats.NewCounter32("store.bigtable.chunk_operations.save_ok")
+	// metric store.bigtable.chunk_operations.save_fail is counter of failed saves
+	chunkSaveFail = stats.NewCounter32("store.bigtable.chunk_operations.save_fail")
+	// metric store.bigtable.chunk_size.at_save is the sizes of chunks seen when saving them
+	chunkSizeAtSave = stats.NewMeter32("store.bigtable.chunk_size.at_save", true)
+	// metric store.bigtable.chunk_size.at_load is the sizes of chunks seen when loading them
+	chunkSizeAtLoad = stats.NewMeter32("store.bigtable.chunk_size.at_load", true)
+	// metric store.bigtable.byte_size.at_save is the number of chunk bytes saved in each bulkApply
+	chunkBytesAtSave = stats.NewMeter32("store.bigtable.chunk_bytes.at_save", true)
+)
+
+func formatRowKey(key schema.AMKey, ts uint32) string {
+	return fmt.Sprintf("%s_%d", key.MKey.String(), ts/Month_sec)
+}
+
+func formatFamily(ttl uint32) string {
+	return fmt.Sprintf("ret_%d", ttlToHours(ttl))
+}
+
+func PrepareChunkData(span uint32, data []byte) []byte {
+	chunkSizeAtSave.Value(len(data))
+	version := chunk.FormatStandardGoTszWithSpan
+	buf := new(bytes.Buffer)
+	binary.Write(buf, binary.LittleEndian, version)
+
+	spanCode, ok := chunk.RevChunkSpans[span]
+	if !ok {
+		// it's probably better to panic than to persist the chunk with a wrong length
+		panic(fmt.Sprintf("Chunk span invalid: %d", span))
+	}
+	binary.Write(buf, binary.LittleEndian, spanCode)
+	buf.Write(data)
+	return buf.Bytes()
+}
+
+func mutationFromWriteRequest(cwr *mdata.ChunkWriteRequest) (*bigtable.Mutation, int) {
+	mut := bigtable.NewMutation()
+	family := formatFamily(cwr.TTL)
+	column := "raw"
+	parts := strings.SplitN(cwr.Key.String(), "_", 2)
+	if len(parts) > 1 {
+		column = parts[1]
+	}
+	value := PrepareChunkData(cwr.Span, cwr.Chunk.Series.Bytes())
+	chunkSizeAtSave.Value(len(value))
+	mut.Set(family, column, bigtable.Timestamp(int64(cwr.Chunk.T0)*1e6), value)
+	return mut, len(value)
+}
+
+func ttlToHours(ttl uint32) uint32 {
+	hours := ttl / (60 * 60)
+	if hours < 1 {
+		hours = 1
+	}
+	return hours
+}
+
+type Store struct {
+	tbl              *bigtable.Table
+	client           *bigtable.Client
+	writeQueues      []chan *mdata.ChunkWriteRequest
+	writeQueueMeters []*stats.Range32
+	readQueue        chan struct{}
+	shutdown         chan struct{}
+	wg               sync.WaitGroup
+	tracer           opentracing.Tracer
+	cfg              *StoreConfig
+}
+
+func NewStore(cfg *StoreConfig, ttls []uint32) (*Store, error) {
+	if cfg.WriteMaxFlushSize >= cfg.WriteQueueSize {
+		return nil, fmt.Errorf("btStore: write-queue-size must be larger then write-max-flush-size")
```

Comment:
config validation -> ConfigProcess()

Reply:
no.   There is no guarantee that users call ConfigProcess() before calling NewStore.

Reply:
* MT input plugins are only meant to be called from MT code base
* we manage the MT code base so we can guarantee whatever we want
* it's our convention to call ConfigProcess to validate configuration before instantiating structures. While it's not the most elegant, it a good practice to keep these steps separate (so we can bail out early upon invalid configuration), when we transition to a better config system I imagine we'll still want separate config validation and object instantiation steps. And sticking to this practice consistently now will also help with migrating later.